### ISSUE TITLE
DatatypeFactory instance caching

### ISSUE THREAD

**Issue by xShadov** [mapstruct/mapstruct#1267]
I noticed that generated code always calls the static factory:

`return DatatypeFactory.newInstance().newXMLGregorianCalendar( c );`

From what I know the creation of the DatatypeFactory is really expensive and could be cached, is that possible? For an app which uses a lot of conversions between XMLGregorianCalendar and other dates the overhead will grow a lot.

**Comment by filiphr** [comment]
Making it cacheable could complicate things quite a bit (if we want the generated code to look good).

What you can do is to create your own mapper that would cache the factory and use that one.

For example:

```java
public class CacheableFactoryMapper {

    private final DateTypeFactory factory;

    public CacheableFactoryMapper() {
        this. factory = DatetypeFactory.newInstance();
    }

    public XMLGregorianCalendar fromDate(Date date) {
        if (date == null) {
            return null;
        }

        GregorianCalendar c = new GregorianCalendar(); 
        c.setTime( date );

        return factory.newXMLGregorianCalendar(c);
    }
}
```

Then in your mappers you will need to add that you use this factory.

```java
@Mapper(uses = {CacheableFactoryMapper.class}) // can be done via @MapperConfig as well
public interface MyMapper {

    // your mappings
}
```

The caching can be done in whichever way you want. It doesn't have to be like I wrote it. You can also do it statically. The `fromDate` can also be a `static` method and MapStruct will invoke it without instantiating the class.

**Comment by sjaakd** [comment]
Solved with #1532
